Doc 12259 vector search c

ElliotFrancisHunter opened the pull request:
Preview link: https://preview.docs-test.couchbase.com/Elvec3/home/index.html
Credentials: https://couchbasecloud.atlassian.net/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords

This PR contains documentation plus release notes for the 3.2.0-beta.3 release of Vector Search for the C platform.

ElliotFrancisHunter committed:
DOC-12259: Relnotes

Added c release notes
Also fly by formatting fix to match 3.1 relnotes format

ElliotFrancisHunter committed:
DOC-12259: Beta 3 updates

Updated what's new page, the nav on C platform to match other updated platforms and the antora to match beta-3.

ElliotFrancisHunter committed:
DOC-12259: Vector Search for C v1

Still needs installation instructions updated and clearly positioned.

Change list:
-- Added copy for vector search and working with vector search plus using C snippets
-- Updated downloads page to correctly point to 3.2.0
-- Updated nav to point to beta API references page
Fly by fixes:
-- Removed dead comments
-- Fixed formatting issue in snippets file

pasin commented on modules/c/partials/release-notes/couchbase-mobile-c-release-note.3.2.0.adoc:
```diff
@@ -35,7 +149,9 @@ None for this release
 ===  Known Issues
 
-None for this release
+* https://issues.couchbase.com/browse/CBL-5551[CBL-5551 - Index trained regardless of warning when using PQ with hihg subquantizer]
```
This issue has been already fixed in beta-3.

ElliotFrancisHunter replied:
Awesome thank you!

ElliotFrancisHunter committed:
Doc-12259: Add missing parameter

Fly by fix for DOC-12258

ElliotFrancisHunter committed:
DOC-12259: V2 vector search doc

Change list:
-- updated snippet copy
-- add vector search attributes to the module attributes file
-- Remove fixed known issue
-- Updated supported API version
-- Added install instructions for supported platforms.

ElliotFrancisHunter committed:
DOC-12259: Add prereq admonition

pasin commented on modules/c/pages/gs-install.adoc:
```diff
@@ -229,11 +240,15 @@ sudo apt install ./{release-dir-dev-ee}
+
+Before you can use Vector Search, you must link:https://www.couchbase.com/downloads/?family=couchbase-lite#LiteExtensions[Install the Vector Search extension].
+The Vector Search extension for the C platform ships with supported prebuilt libraries containing the required dependencies.
+You need to set the `LD_LIBRARIES_PATH` to the extension location instead of installing the libraries yourself.
```
`to the supported prebuilt libraries location`?

pasin replied:
> Before you can use Vector Search, you must link:https://www.couchbase.com/downloads/?family=couchbase-lite#LiteExtensions[Install the Vector Search extension].

Before you can use Vector Search, you must download and install the vector search library to the location in your project where the vector search library can be accessed and loaded while the executable is running. 

In the code, before opening the database and using the vector search extension, you must call `CBL_SetExtensionPath` function to set the path to the installed location of the vector search library.

Noted : The step to call CBL_SetExtensionPath is the same for all platforms for CBL-C. There should be a codesnipt for this.

ElliotFrancisHunter replied:
Done!

pasin commented on modules/c/pages/gs-install.adoc:
```diff
@@ -335,6 +445,62 @@ In this instance the release comprises a _ready to adapt_ application project.
 
 Once the install is finished, you can build and run this skeleton app.
 
+=== Install Vector Search for Android
```
This is complicated. 

1. 'com.couchbase.lite:couchbase-lite-android-ee:{version-maintenance-android}'` This is CBL-Android, not CBL-C. CBL-C provides native shared library that developers can use to write JNI codes for their application. So CBL-C developers for Android platform will not use `couchbase-lite-android-ee` library.

2. `com.couchbase.lite:couchbase-lite-java-vector-search-arm64` is pure Java so it's not relevant here.

3. `couchbase-lite-android-vector-search-arm64 / x86_64` : This is interesting as users can just use this instead of the native android libraries provided by CBL-C. However, CBL-C 3.2.0 beta is not compatible with `couchbase-lite-java-android-search-arm64/x86_64` beta-2 so the developers cannot use the current `couchbase-lite-java-android-search`. So it would be best that we don't include this instruction here to avoid confusion. But this would work for GA.

4. CBL-C Vector Search provides a downloable file that only contains the vector search libraries that the developer will need to include them in their app. The steps to include the prebuilt native library can be found here: https://developer.android.com/studio/projects/gradle-external-native-builds.

If this is not clear, we can sync up about this tomorrow.

pasin commented on modules/c/pages/gs-install.adoc:
```diff
@@ -257,6 +272,12 @@ Be sure to select x64 for 64-bit builds
+To use the Vector Search extension:
+
+. Download and extract the link:https://www.couchbase.com/downloads/?family=couchbase-lite#LiteExtensions[Vector Search extension].
+. Put the library in your development environment.
```
CBL provides CouchbaseLiteVectorSearch.dll. Similar to the instruction to install CBL for windows above, 

```
Couchbase Lite Vector Search does not have any preferred installation path for the `.dll`. It is up to you to determine where best to place it so it is available during execution.
```

ElliotFrancisHunter replied:
Done!

pasin commented on modules/c/pages/gs-install.adoc:
```diff
@@ -289,6 +310,51 @@ Then, within the dialog:
+[{tabs}]
+=====
+
+Objective-C::
```
This step is for iOS app not macOS so we could remove it.

ElliotFrancisHunter replied:
Removed!

pasin commented on modules/c/pages/gs-install.adoc:
```diff
@@ -289,6 +310,51 @@ Then, within the dialog:
+. Download and extract the link:https://www.couchbase.com/downloads/?family=couchbase-lite#LiteExtensions[Vector Search extension] into your XCode project location.
+. Select your target settings in XCode and drag *CouchbaseLiteVectorSearch.xcframework* from your Finder to the *Frameworks, Libraries, and Embedded Content* section.
+. Import the xcframework and start using it in your project.
+. Find the extension path using the appropriate snippet:
```
The step is similar to the step to install CBL dylib but the differences are :

1. We don't support homebrew. We only package the zip file.

2. It will be CouchbaseLiteVectorSearch.dylib.

3. Developer must download and install the vector search library to the location in your project where the vector search library can be accessed and loaded while the executable is running. (Same as the linux platform - it will depend on the developer where to put the vector search library for the application.).

ElliotFrancisHunter replied:
I think I've done this

pasin reviewed (commented):
I will have to review the release notes. Will try to do it tomorrow.

ElliotFrancisHunter committed:
DOC-12259: Apply feedback - V3

pasin commented on modules/c/pages/gs-install.adoc:
```diff
@@ -257,6 +282,14 @@ Be sure to select x64 for 64-bit builds
+. Download and extract the link:https://www.couchbase.com/downloads/?family=couchbase-lite#LiteExtensions[Vector Search extension].
+. Put the library in your development environment.
+
+NOTE: Couchbase Lite Vector Search does not have any preferred installation path for the `.dll`. It is up to you to determine where best to place it so it is available during execution.
```
Not sure how you would like to organzie this but same as the other platforms, users will need to set the extension path as well.

ElliotFrancisHunter replied:
I think I'm going to repeat this rather than have it in the summary to ensure readers won't miss it when looking at their specific platform.

pasin commented on modules/c/pages/gs-install.adoc:
```diff
@@ -289,12 +322,26 @@ Then, within the dialog:
 
+----
+
+NOTE: Couchbase Lite Vector Search does not have any preferred installation path for the `.dll`. It is up to you to determine where best to place it so it is available during execution.
```
This NOTE is for windows.

ElliotFrancisHunter replied:
Removed!

pasin commented on modules/c/pages/gs-install.adoc:
```diff
@@ -335,6 +426,57 @@ In this instance the release comprises a _ready to adapt_ application project.
+----
+
+--
+=====
```
Same as the other platforms, users will need to set the extension path as well.

ElliotFrancisHunter replied:
Done!

ElliotFrancisHunter committed:
DOC-12259: V4 Feedback update

Change list:
-- Remove incorrect admonitions on install page
-- Add missing snippet to install steps
-- Remove a dupe ticket and private tickets from C release notes

ElliotFrancisHunter commented:
@pasin there seems to be an error with the VectorSearch.h snippets file preventing the build from passing. Could you have a look please?

pasin committed:
Fix PR validation error

Using build number to 43 (CBL-C 3.2.0-Beta.3) as there is a small change to the CBLPredictiveModel API in the newer build for GA.

pasin commented on modules/c/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,230 @@
+|No default
+|A SQL++ expression indicating where to get the vectors. 
+A document property for embedded vectors
+`prediction()` to call a registered Predictive model.
```
"or" is missing. `for embedded vectors or prediction()`

ElliotFrancisHunter replied:
Done

pasin commented on modules/c/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,230 @@
+|Expression Language
+|Required
+|No default
+|Specifies the query language used. This can be SQL++ or N1Q1 query.
```
This can be N1QL or JSON query.

For the JSON query syntax, we can link to https://github.com/couchbase/couchbase-lite-core/wiki/JSON-Query-Schema.

ElliotFrancisHunter replied:
Done!

pasin commented on modules/c/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,230 @@
+
+. First, create the standard configuration, setting up an expressions, number of dimensions and number of centroids for the vector embedding.
+
+. Next, create a collection variable passing it the contents of a given collection from our `database` object, in our case, the `colors` collection.
```
Next, get the `colors` collection from the database.

ElliotFrancisHunter replied:
Much more to the point, nice! Changed

pasin commented on modules/c/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,230 @@
+
+. Next, create a collection variable passing it the contents of a given collection from our `database` object, in our case, the `colors` collection.
+
+. Finally, create a vector index, `colors_index`, on a collection and pass it our configuration.
```
Finally, create a vector index, `colors_index`, on the collection with the configuration.

ElliotFrancisHunter replied:
Done!

pasin reviewed (dismissed):
Found a few small things while reviewing some content with @rajagp.

pasin committed:
Fix iOS PR validation

ElliotFrancisHunter committed:
DOC-12259: V4 feedback changes

Change list:
-- Typo fixes in working with pages
-- Tighten up snippet copy for the color examples

ElliotFrancisHunter committed:
DOC-12259: Vector Search install updates and formatting fixes

osfameron commented on modules/c/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,231 @@
+NOTE: You can use less storage by using the `prediction()` function as the encoded vectors will only be stored in the index. 
+However, the index time will be longer as vector embedding generation is occurring at run time.
+
+== Vector SQL++ Functions
```
Check this renders properly, if not you may need to use `{sqlpp}`

ElliotFrancisHunter replied:
Yeah it's all fine

osfameron commented on modules/ROOT/pages/cbl-whatsnew.adoc:
```diff
@@ -17,6 +17,27 @@ NOTE: Couchbase Lite 3.0 introduces some breaking changes. +
 If you are upgrading from 2.x, please refer to the appropriate upgrade page -- see: <<lbl-upgrade>>
 Users should be able to upgrade to 3.1.x from 3.0.x without manual intervention.
 
+== Release 3.2.0 - Beta 3 - C (June 2024)
```
Just to check I understand what you're doing here: all this info is in the whatsnew during beta, and referred to from the minimal relnotes.

After beta, will that be folded back into the GA relnote?

ElliotFrancisHunter replied:
Most likely I'll fold them into a general description of Vector Search plus the API additions yeah.
